docs/doxygen/collectives.md

```diff
@@ -1,20 +1,150 @@
+enumeration ucc\_coll\_type\_t. The semantics are briefly described here,
+however in most cases it agrees with the semantics of collective operations in
+the popular programming models such as MPI and OpenSHMEM. When they differ, the
+semantics changes are documented. All collective operations operate on the team.
+For the collective operations defined in the enumeration, all participants of
+the team participate in the collective operations. Further the team should be
+created with endpoints, where the “eps” should be ordered and contiguous.
 
-\b Reduction types: The reduction operation supported by UCC\_{ALLREDUCE, REDUCE} operation is defined by the enumeration ucc\_op\_t. The valid datatypes for the reduction is defined by the enumeration ucc\_datatype\_t.
+UCC supports three types of collective operations: (a) UCC\_{ALLTOALL,ALLTOALLV,
+ALLGATHER, ALLGATHERV, ALLREDUCE, REDUCE_SCATTER,REDUCE_SCATTERV, BARRIER}
+operations where all participants contribute to the results and receive the
+results (b) UCC\_{REDUCE, GATHER, GATHERV, FANIN} where all participants
+contribute to the result and one participant receives the result. The
+participant receiving the result is designated as root. (c) UCC\_{BROADCAST,
+SCATTER, SCATTERV, FANOUT} where one participant contributes to the result, and
+all participants receive the result. The participant contributing to the result
+is designated as root.
 
-\b Ordering: The team can be configured for ordered collective operations or unordered collective operations. For unordered collectives, the user is required to provide the “tag”, which is an unsigned 64-bit integer. 
++ The UCC\_COLL\_TYPE\_BCAST operation moves the data from the src buffer of
+root participant to the dest buffer of all participants.
 
-\b Synchronized and Non-Synchronized Collectives: In the synchronized collective model, on entry, the participants cannot read or write to other participants without ensuring all participants have entered the collective operation. On the exit of the collective operation, the participants may exit after all participants have completed the reading or writing to the buffers.
++ The UCC\_COLL\_TYPE\_BARRIER synchronizes all participants of the collective
+operation. In this routine, first, each participant waits for all other
+participants to enter the operation. Then, once it learns the entry of all other
+participants into the operation, it exits the operation completing it locally.
 
-In the non-synchronized collective model, on entry, the participants can read or write to other participants. If the input and output buffers are defined on the team and RMA operations are used for data transfer, it is the responsibility of the user to ensure the readiness of the buffer. On exit, the participants may exit once the read and write to the local buffers are completed. 
++ The UCC\_COLL\_TYPE\_FANIN gathers the data from all participants in the team
```

Comment:
FANIN/FANOUT don't move data. These are essentially synchronizing primitives (like UCC_COLL_TYPE_BaRRIER). Fanin can be viewed as 0 byte reduce/gather, and fanout is 0 byte bcast/scatter.

Reply:
I see. Small change to what you are saying - we don't have to write it as a 0-byte msg. Probably, we could write it as sync msg and leave it to the implementation, if it wants to use 0-byte or more that is up to the implementation.

I will rewrite the semantics. Let's see if it will capture the use cases.